himbaechel/uarch/gowin/pack.cc

```diff
@@ -3072,6 +3076,89 @@ struct GowinPacker
         }
     }
 
+    // =========================================
+    // Enable UserFlash
+    // =========================================
+    void pack_userflash()
+    {
+        log_info("Pack UserFlash cells...\n");
+
+        for (auto &cell : ctx->cells) {
+            auto &ci = *cell.second;
+            if (!is_userflash(&ci)) {
+                continue;
+            }
+
+            if (ci.type.in(id_FLASH96K, id_FLASH256K, id_FLASH608K)) {
+                // enable
+                ci.addInput(id_INUSEN);
+                ci.connectPort(id_INUSEN, ctx->nets.at(ctx->id("$PACKER_GND")).get());
+            }
+            // rename ports
+            for (int i = 0; i < 32; ++i) {
+                ci.renamePort(ctx->idf("DIN[%d]", i), ctx->idf("DIN%d", i));
+                ci.renamePort(ctx->idf("DOUT[%d]", i), ctx->idf("DOUT%d", i));
+            }
+            if (ci.type.in(id_FLASH96K)) {
+                for (int i = 0; i < 6; ++i) {
+                    ci.renamePort(ctx->idf("RA[%d]", i), ctx->idf("RA%d", i));
+                    ci.renamePort(ctx->idf("CA[%d]", i), ctx->idf("CA%d", i));
+                    ci.renamePort(ctx->idf("PA[%d]", i), ctx->idf("PA%d", i));
+                }
+                for (int i = 0; i < 2; ++i) {
+                    ci.renamePort(ctx->idf("MODE[%d]", i), ctx->idf("MODE%d", i));
+                    ci.renamePort(ctx->idf("SEQ[%d]", i), ctx->idf("SEQ%d", i));
+                    ci.renamePort(ctx->idf("RMODE[%d]", i), ctx->idf("RMODE%d", i));
+                    ci.renamePort(ctx->idf("WMODE[%d]", i), ctx->idf("WMODE%d", i));
+                    ci.renamePort(ctx->idf("RBYTESEL[%d]", i), ctx->idf("RBYTESEL%d", i));
+                    ci.renamePort(ctx->idf("WBYTESEL[%d]", i), ctx->idf("WBYTESEL%d", i));
+                }
+            } else {
+                for (int i = 0; i < 9; ++i) {
+                    ci.renamePort(ctx->idf("XADR[%d]", i), ctx->idf("XADR%d", i));
+                }
+                for (int i = 0; i < 6; ++i) {
+                    ci.renamePort(ctx->idf("YADR[%d]", i), ctx->idf("YADR%d", i));
+                }
+            }
+            // add invertor
+            int lut_idx = 0;
+            auto add_inv = [&](IdString port, PortType port_type) {
+                if (!port_used(&ci, port)) {
+                    return;
+                }
+
+                CellInfo *lut = ctx->createCell(create_aux_name(ci.name, lut_idx, "_lut$"), id_LUT4);
```

Comment:
creating a cell while also iterating over the cells map is not permitted - usually the best thing to do is build up a vector of user flash cells and then iterate over that to do the actual task of packing with creating cells, etc

Reply:
My mistake. How's that?